### PR TITLE
Airport-Challenge-CatherineD

### PR DESCRIPTION
**Pull request by hsdcjd**

### CODE REVIEW

File: airport-challenge-solution.md [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+b Need to be able to give the Airport a Plane
+c LandedPlanes list length should increase by 1 when a Plane is added to the empty LandedPlanes list
+
+**Test 1** - `landedPlanes` length increases by 1 when `landPlane` is called with _instance of `Plane`_
+
+**Test 2** - `landPlane` should only add `Plane` instances to the `landedPlanes`
+
+**Test 3** - edge Case - falsy values should not be added to the Airport
+
+//I am not clear if if Test 2 and 3 are the same explicitly?
```
Comment (hsdcjd) [inline airport-challenge-solution.md:31]: I understand that test 2 lands a plane, but test 3 needs to exist to check if adding something else is possible, for which I tried a string and an integer.

File: airport-challenge-solution.md [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+|         |                             | newLandedCapacity(@Integer) | @Void    |
+|         |                             | getLandedPlaneCount()       | @Integer |
+| Plane   | id @String                  | getId()                     | @String  |
+
+## Tests
+
+a Need to check if is AirportFull True prevents Plane going into 'landedPlanes'
+b Need to check if is AirportFull False allows Plane to go into 'landedPlanes'
+//this approach requires the caller to ask if airport is full first (then do something), whereas would be better to have prevention of landPlane if full
+
```
Comment (hsdcjd) [inline airport-challenge-solution.md:78]: Did not use isAirportFull in the end (Boolean return), as this would have required another method anyway after the return from that.

File: airport-challenge-solution.md [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+|         |                             | takeOffPlane(@Plane)        | @Void    |
+|         |                             | isPlanePresent(@Plane)      | @Boolean |
+| Plane   | id @String                  | getId()                     | @String  |
+
+## Tests 9&10
+
+9. try to take off a plane(id) which is not in the airport (landedPlanes)
+10. try land a second instance of a plane(id) that is already in landedPlanes
+    (need to add a Boolean return to landPlane function?, yes but did not end up using)
+
```
Comment (hsdcjd) [inline airport-challenge-solution.md:130]: Test numbering/presentation gets messy - variation throughout.

File: airport-challenge-solution.md [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+I want to prevent takeoff when weather is stormy //not attempted
+
+As an air traffic controller
+To ensure safety
+I want to prevent landing when weather is stormy //not attempted
+
+As an air traffic controller
+To count planes easily
+Planes that have landed must be at an airport //this is covered within the Airport Class, landedPlanes array.
+```
```
Comment (hsdcjd) [inline airport-challenge-solution.md:158]: Read ahead to extended criteria to understand if I am attempting them within my earlier tests.

File: airport-challenge-solution.md [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+
+9. try to take off a plane(id) which is not in the airport (landedPlanes)
+10. try land a second instance of a plane(id) that is already in landedPlanes
+    (need to add a Boolean return to landPlane function?, yes but did not end up using)
+
+## Testing Edge Cases
+
+Have already shown it is not possible to land something that is not a plane - a string and a value, in test 3.
+Have also shown that a plane that is already landed can not land a second time, in test 10.
+
```
Comment (hsdcjd) [inline airport-challenge-solution.md:135]: Read ahead to look at explanations for Edge cases, to understand if I am attempting them as part of the basic criteria. Also need to understand Edge cases more anyway.

File: src/airport.js [hunk 6]
```diff
@@ -1,6 +1,48 @@
+const Plane = require('./Plane');
+
 class Airport {
-  // here's a starting point for you
+  landedPlanes = [];
+  landedCapacity;
+
+  // constructor(defaultLandedCapacity) {
+  //   this.landedCapacity = defaultLandedCapacity;
+  // }
```
Comment (hsdcjd) [inline src/airport.js:10]: This was the original constructor before later adding a rejection of negative values and floats. I don't think I could have done that earlier, as it was an edge case which I didn't consider until the end. But in the future I may have that foresight.

File: src/airport.js [hunk 7]
```diff
@@ -1,6 +1,48 @@
...
+      this.landedCapacity = 0;
+    }
+  }
+  landPlane = plane => {
+    if (plane instanceof Plane && this.getLandedPlaneCount() < this.getLandedCapacity() && !this.isPlanePresent(plane)) {
+      this.landedPlanes.push(plane);
+      // return true;
+      //   } else {
+      //       return false;
+    }
```
Comment (hsdcjd) [inline src/airport.js:24]: Had originally returned this a Boolean, however it was unused, and I checked the second landing of a plane (which came in a late test) by using a then existing method getPlaneCount

File: src/airport.js [hunk 8]
```diff
@@ -1,6 +1,48 @@
...
+  isPlanePresent(plane) {
+    const indexOfPlaneInLandedPlanes = this.landedPlanes.findIndex(landedPlane => landedPlane.id === plane.id);
+    return indexOfPlaneInLandedPlanes > -1;
+  }
+  takeOffPlane = plane => {
+    const indexOfPlaneInLandedPlanes = this.landedPlanes.findIndex(landedPlane => landedPlane.id === plane.id);
+    if (indexOfPlaneInLandedPlanes > -1) {
+      this.landedPlanes.splice(indexOfPlaneInLandedPlanes, 1);
+       } return (indexOfPlaneInLandedPlanes > -1) ? true : false;
+    }
```
Comment (hsdcjd) [inline src/airport.js:44]: I could also have taken the approach here (re-takeOffPlane) used in the re-landing of Plane above. And not used a Boolean return but used a previous existing method such as getLandedPlaneCount

File: src/plane.js [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+
+module.exports = Plane;
+
+
+// Test to check that planes cannot take off when they are not at the airport
+// const planeStatus = 'notAtAirport';
+
+// if (planeStatus === 'notAtAirport') {
+//   assert.strictEqual(takeOff(), false, 'Cannot take off when plane is not at the airport')
+// }
```
Comment (hsdcjd) [inline src/plane.js:19]: Had googled this but not used. Was thinking of putting plane status into the Plane Class. Happy with my approach in the airport.

File: test/airport.spec.js [hunk 10]
```diff
@@ -0,0 +1,331 @@
...
+plane1 = new Plane('Delta');
+expected = false;
+
+// Act - this is running the methods to be tested and collecting the actual results
+airport.landPlane(plane1); //land a plane at the airport first!
+//console.log(airport.isPlanePresent(plane1));
+airport.takeOffPlane(plane1);
+//console.log(airport.isPlanePresent(plane1));
+airport.takeOffPlane(plane1); 
+actual = airport.takeOffPlane(plane1);
```
Comment (hsdcjd) [inline test/airport.spec.js:292]: I had to play with this one a bit. Should I remove all commented-out console.logs?

File: test/airport.spec.js [hunk 11]
```diff
@@ -0,0 +1,331 @@
...
+actual = undefined;
+result = undefined;
+
+console.log(`============================`);
+
+//Test 9 - try to take off a plane(id) which is not in the airport (landedPlanes)
+
+console.log(`============================`);
+console.log(`Test 9 - try to takeOffPlane Plane(id) that is not in landedPlanes`);
+
```
Comment (hsdcjd) [inline test/airport.spec.js:225]: Wend round in circles and revisited test 9 a few times. Was making it more complicated than it needed to be. Went with a Boolean return in takeOffPlane method in the end, but would rather have a 'throw error'.

File: test/airport.spec.js (hunk 10 above)
Comment (hsdcjd) [inline test/airport.spec.js:292]: Was thinking to use getPlaneCount, but this doesn't work if already gone to zero anyway.

File: test/airport.spec.js [hunk 12]
```diff
@@ -0,0 +1,331 @@
...
+//Test 12 - test to check that only a positive integer is accepted for capacity values
+
+console.log(`============================`);
+console.log(`Test 12 - 'try to set defaultAirportCapacity as a float or negative'`);
+
+// Arrange - this is setting up the variables required for the test
+defaultLandedCapacity = -4; //6.5 also tried
+airport = new Airport(defaultLandedCapacity);
+expected = 0; //would ideally throw an error but not sure how to do that
+
```
Comment (hsdcjd) [inline test/airport.spec.js:316]: defaultAirportCapacity becomes zero if a -ve or float is attempted, but would rather have a 'throw error'

**Review by hsdcjd — COMMENTED** [review]
Comments are throughout airport-challenge-solution.md file, src/airport.js, src/plane.js and test/airport.spec.js
